File: src/hammond/services/greeter.py
import random

import restate

from datetime import timedelta
from pydantic import BaseModel


# You can also just use a typed dict, without Pydantic
class GreetingRequest(BaseModel):
    name: str


class Greeting(BaseModel):
    message: str


greeter = restate.Service("Greeter")


@greeter.handler()
async def greet(ctx: restate.Context, req: GreetingRequest) -> Greeting:
    # Durably execute a set of steps; resilient against failures
    greeting_id = str(ctx.uuid())
    await ctx.run_typed(
        "notification", send_notification, greeting_id=greeting_id, name=req.name
    )
    await ctx.sleep(timedelta(seconds=1))
    await ctx.run_typed(
        "reminder", send_reminder, greeting_id=greeting_id, name=req.name
    )

    # Respond to caller
    return Greeting(message=f"You said hi to {req.name}!")


def send_notification(greeting_id: str, name: str):
    if random.random() < 0.7 and name == "Alice":  # 70% chance of failure
        print(f"[👻 SIMULATED] Failed to send notification: {greeting_id} - {name}")
        raise Exception(
            f"[👻 SIMULATED] Failed to send notification: {greeting_id} - {name}"
        )
    print(f"Notification sent: {greeting_id} - {name}")


def send_reminder(greeting_id: str, name: str):
    if random.random() < 0.7 and name == "Alice":  # 70% chance of failure
        print(f"[👻 SIMULATED] Failed to send reminder: {greeting_id}")
        raise Exception(f"[👻 SIMULATED] Failed to send reminder: {greeting_id}")
    print(f"Reminder sent: {greeting_id}")

File: greeter.go
package main

import (
	"fmt"
	"math/rand"
	"time"

	restate "github.com/restatedev/sdk-go"
)

// Greeter is a struct which represents a Restate service; reflection will turn exported methods into service handlers
type Greeter struct{}

func (Greeter) Greet(ctx restate.Context, name string) (string, error) {
	//  Durably execute a set of steps; resilient against failures
	greetingId := restate.UUID(ctx).String()

	if _, err := restate.Run(ctx,
		func(ctx restate.RunContext) (restate.Void, error) {
			return SendNotification(greetingId, name)
		},
		restate.WithName("Notification"),
	); err != nil {
		return "", err
	}

	if err := restate.Sleep(ctx, 1*time.Second); err != nil {
		return "", err
	}

	if _, err := restate.Run(ctx,
		func(ctx restate.RunContext) (restate.Void, error) {
			return SendReminder(greetingId, name)
		},
		restate.WithName("Reminder"),
	); err != nil {
		return "", err
	}

	// Respond to caller
	return "You said hi to " + name + "!", nil
}

func SendNotification(greetingId string, name string) (restate.Void, error) {
	if rand.Float32() < 0.7 && name == "Alice" { // 70% chance of failure
		fmt.Printf("[👻 SIMULATED] Failed to send notification: %s - %s\n", greetingId, name)
		return restate.Void{}, fmt.Errorf("[👻 SIMULATED] Failed to send notification: %s - %s", greetingId, name)
	}
	fmt.Printf("Notification sent: %s - %s\n", greetingId, name)
	return restate.Void{}, nil
}

func SendReminder(greetingId string, name string) (restate.Void, error) {
	if rand.Float32() < 0.7 && name == "Alice" { // 70% chance of failure
		fmt.Printf("[👻 SIMULATED] Failed to send reminder: %s\n", greetingId)
		return restate.Void{}, fmt.Errorf(" [👻 SIMULATED] Failed to send reminder: %s", greetingId)
	}
	fmt.Printf("Reminder sent: %s\n", greetingId)
	return restate.Void{}, nil
}
